File: app/src/main/java/io/github/benoitduffez/cupsprint/detect/Merger.java
package io.github.benoitduffez.cupsprint.detect;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.github.benoitduffez.cupsprint.L;

class Merger {
    void merge(List<PrinterRec> httpRecs, List<PrinterRec> httpsRecs) {
        List<PrinterRec> tmpRecs = new ArrayList<>();
        for (PrinterRec httpRec : httpRecs) {
            boolean match = false;
            for (PrinterRec httpsRec : httpsRecs) {
                try {
                    if (httpRec.getQueue().equals(httpsRec.getQueue()) &&
                            httpRec.getHost().equals(httpsRec.getHost()) &&
                            httpRec.getPort() == httpsRec.getPort()) {
                        match = true;
                        break;
                    }
                } catch (Exception e) {
                    L.e("Invalid record in merge", e);
                }
            }
            if (!match) {
                tmpRecs.add(httpRec);
            }
        }
        for (PrinterRec rec : tmpRecs) {
            httpsRecs.add(rec);
        }
        Collections.sort(httpsRecs);
    }
}

File: app/src/main/java/io/github/benoitduffez/cupsprint/detect/Merger.kt
package io.github.benoitduffez.cupsprint.detect

import timber.log.Timber
import java.util.*

internal class Merger {
    fun merge(httpRecs: List<PrinterRec>, httpsRecs: MutableList<PrinterRec>) {
        val tmpRecs = ArrayList<PrinterRec>()
        for (httpRec in httpRecs) {
            var match = false
            for (httpsRec in httpsRecs) {
                try {
                    if (httpRec.queue == httpsRec.queue &&
                            httpRec.host == httpsRec.host &&
                            httpRec.port == httpsRec.port) {
                        match = true
                        break
                    }
                } catch (e: Exception) {
                    Timber.e(e, "Invalid record in merge")
                }
            }
            if (!match) {
                tmpRecs.add(httpRec)
            }
        }
        for (rec in tmpRecs) {
            httpsRecs.add(rec)
        }
        httpsRecs.sort()
    }
}
